src/migration/service-agreements-op-database-migration.js

```diff
@@ -0,0 +1,137 @@
+/* eslint-disable no-await-in-loop */
+import BaseMigration from './base-migration.js';
+
+let wrongAgreementsCount = 0;
+const MAX_BATCH_SIZE = 10000;
+const CONCURRENCY = 3;
+
+class ServiceAgreementsOpDatabaseMigration extends BaseMigration {
+    constructor(
+        migrationName,
+        logger,
+        config,
+        blockchainModuleManager,
+        repositoryModuleManager,
+        serviceAgreementService,
+        ualService,
+    ) {
+        super(migrationName, logger, config);
+        this.blockchainModuleManager = blockchainModuleManager;
+        this.repositoryModuleManager = repositoryModuleManager;
+        this.serviceAgreementService = serviceAgreementService;
+        this.ualService = ualService;
+    }
+
+    async executeMigration() {
+        let migrationInfo = await this.getMigrationInfo();
+        if (!migrationInfo?.lastProcessedTokenId) {
+            migrationInfo = {
+                lastProcessedTokenId: 0,
+            };
+        }
+
+        const numberOfActiveServiceAgreements =
+            await this.repositoryModuleManager.getNumberOfActiveServiceAgreements();
+        let processed = 0;
+        const batchSize =
+            numberOfActiveServiceAgreements > MAX_BATCH_SIZE
+                ? MAX_BATCH_SIZE
+                : numberOfActiveServiceAgreements;
+
+        while (processed < numberOfActiveServiceAgreements) {
+            const serviceAgreementsToProcess =
+                await this.repositoryModuleManager.getServiceAgreements(
+                    migrationInfo.lastProcessedTokenId,
+                    batchSize,
+                );
+            let promises = [];
+
+            for (const serviceAgreement of serviceAgreementsToProcess) {
+                promises.push(this.processServiceAgreement(serviceAgreement));
+
+                if (
+                    promises.length >= CONCURRENCY ||
+                    promises.length === serviceAgreementsToProcess.length
+                ) {
+                    try {
+                        await Promise.all(promises);
+                    } catch (error) {
+                        this.logger.warn(
+                            `Unable to process invalid service agreements. Error: ${error}`,
+                        );
+                    }
+                    promises = [];
+                    migrationInfo.lastProcessedTokenId =
+                        serviceAgreementsToProcess.slice(-1)[0].tokenId;
```

Comment:
We're not sorting agreements by tokenId when getting them from the operationaldb